src/Sparrow/Platform/Posix/SmapsReader.cs

```diff
@@ -0,0 +1,357 @@
+        {
+            var read = fileStream.Read(_smapsBuffer[bufferIndex], 0, _smapsBuffer[bufferIndex].Length);
+            _endOfBuffer[bufferIndex] = read;
+            return read;
+        }
+
+        public (long Rss, long SharedClean, long PrivateClean, DynamicJsonArray Json) CalculateMemUsageFromSmaps()
+        {
+            var state = SearchState.None;
+            var dja = new DynamicJsonArray();
+
+            using (var currentProcess = Process.GetCurrentProcess())
+            using (var fileStream = new FileStream($"/proc/{currentProcess.Id}/smaps", FileMode.Open, FileAccess.Read, FileShare.Read))
+            {
+                var read = ReadFromFile(fileStream, _currentBuffer);
+                var offsetForNextBuffer = 0;
+                long tmpRss = 0, tmpSharedClean = 0, tmpPrivateClean = 0;
+                string resultString = null;
+                long valSize = 0, valRss = 0, valPrivateDirty = 0, valSharedDirty = 0, valSharedClean = 0, valPrivateClean = 0;
+                while (true)
+                {
+                    if (read == 0)
+                    {
+                        return (tmpRss, tmpSharedClean, tmpPrivateClean, dja);
+                    }
+
+                    var switchBuffer = false;
+                    for (var i = offsetForNextBuffer; i < _endOfBuffer[_currentBuffer]; i++)
+                    {
+                        byte[] term;
+                        var offset = 0;
+                        if (_smapsBuffer[_currentBuffer][i] == 'r')
+                            term = _rwsBytes;
+                        else if (_smapsBuffer[_currentBuffer][i] == 'R')
+                            term = _rssBytes;
+                        else if (_smapsBuffer[_currentBuffer][i] == 'S')
+                        {
+                            term = _sizeBytes; // or SharedDirty or SharedCleanBytes (which ARE longer in length from Size)
+                            offset = _sharedCleanBytes.Length - term.Length;
+                        }
+                        else if (_smapsBuffer[_currentBuffer][i] == 'P')
+                        {
+                            term = _privateCleanBytes; // or PrivateDirty (which is not longer in length from PrivateCleanBytes)
+                        }
+                        else if (_smapsBuffer[_currentBuffer][i] == 'L')
+                        {
+                            term = _lockedBytes;
+                        }
+                        else
+                            continue;
+
+                        // check if the current buffer too small for the search term and read next buff if so
+                        if (switchBuffer == false && i + term.Length + offset > _endOfBuffer[_currentBuffer])
+                        {
+                            var nextBuffer = (_currentBuffer + 1) % 2;
+                            var _ = ReadFromFile(fileStream, nextBuffer);
```

Comment:
We actually do care, if this is the case, we'll need to stop reading more data.

Reply:
True. I was relying on next iteration - _endOfBuffer[_currentBuffer] equals 0 and nothing happens. But it seems it would get into infinite loop if that the case.